core/embed/rust/src/micropython/dict.rs

```diff
@@ -0,0 +1,68 @@
+use core::convert::TryFrom;
+
+use crate::error::Error;
+
+use super::{ffi, gc::Gc, map::Map, obj::Obj};
+
+/// Insides of the MicroPython `dict` object.
+pub type Dict = ffi::mp_obj_dict_t;
+
+impl Dict {
+    /// Allocate a new dictionary on the GC heap, empty, but with a capacity of
+    /// `capacity` items.
+    pub fn alloc_with_capacity(capacity: usize) -> Gc<Self> {
+        unsafe {
+            let ptr = ffi::mp_obj_new_dict(capacity);
+            Gc::from_raw(ptr.as_ptr().cast())
+        }
+    }
+
+    /// Constructs a dictionary definition by taking ownership of given [`Map`].
+    pub fn with_map(map: Map) -> Self {
+        unsafe {
+            Self {
+                base: ffi::mp_obj_base_t {
+                    type_: &ffi::mp_type_dict,
+                },
+                map,
+            }
+        }
+    }
+
+    /// Returns a reference to the contained [`Map`].
+    pub fn map(&self) -> &Map {
+        &self.map
+    }
+
+    /// Returns a mutable reference to the contained [`Map`].
+    pub fn map_mut(&mut self) -> &mut Map {
+        &mut self.map
+    }
+}
+
+impl From<Gc<Dict>> for Obj {
+    fn from(value: Gc<Dict>) -> Self {
+        // SAFETY:
+        //  - `value` is an object struct with a base and a type.
+        //  - `value` is GC-allocated.
+        unsafe { Obj::from_ptr(Gc::into_raw(value).cast()) }
+    }
+}
+
+impl TryFrom<Obj> for Gc<Dict> {
+    type Error = Error;
+
+    fn try_from(value: Obj) -> Result<Self, Self::Error> {
+        if unsafe { ffi::mp_type_dict.is_type_of(value) } {
+            // SAFETY: We assume that if `value` is an object pointer with the correct type,
+            // it is always GC-allocated.
```

Comment:
Yes. The comment is correct, the assumption maybe not hard enough to break, but practical enough for now I think. Theoretically we could validate that the pointer lays in the GC-managed range. Thoughts? You already have to be careful with the values you pass in anyway, because of the aliasing requirements mentioned in the Buffer-related discussion. Sounds like a similar task.

Reply:
let's change the wording to "it is GC-allocated or static (see `Gc` for details)", dtto for the `List` case, and add a couple words to `Gc`...